packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js

```diff
@@ -0,0 +1,61 @@
+import { useEffect, useRef, useState } from 'react';
+import { useLocation } from 'react-router-dom';
+import { useApolloClient } from '@apollo/react-hooks';
+
+import addToCache from './addToCache';
+import getRouteComponent from './getRouteComponent';
+
+const IS_LOADING = { isLoading: true };
+
+export const useMagentoRoute = () => {
+    const [componentMap, setComponentMap] = useState(new Map());
+    const { apiBase } = useApolloClient();
+    const { pathname } = useLocation();
+    const isMountedRef = useRef(false);
+
+    const routeData = componentMap.get(pathname);
+
+    // ask Magento for a RootComponent that matches the current pathname
+    useEffect(() => {
+        // avoid asking if we already know the answer
+        const isKnown = componentMap.has(pathname);
+
+        // ask again following a prior failure
+        const isNotFound = isKnown && componentMap.get(pathname).id === -1;
+        const shouldReload = isNotFound && navigator.onLine;
+
+        if (!isKnown || shouldReload) {
+            getRouteComponent(apiBase, pathname).then(
+                ({ component, id, pathname, routeError }) => {
+                    // avoid setting state if unmounted
+                    if (!isMountedRef.current) {
+                        return;
+                    }
+
+                    // add the pathname to the browser cache
+                    addToCache(pathname);
```

Comment:
In the old implementation we had a `catch` that logged errors in development. Can we add that here? I know we're probably changing this soon so nbd either way.

Reply:
I believe that is [here](https://github.com/magento/pwa-studio/pull/1966/files#diff-062ef3510d7860a80e69ce915563639aR49) still. One of the other reviewers suggested that I remove the `process.env` check since it was redundant, so I did that.

Reply:
Cool, didn't realize we eliminated consoles in prod.